lib/Transforms/LowerPolynomialEval/Patterns.cpp

```diff
@@ -0,0 +1,201 @@
+#include "lib/Transforms/LowerPolynomialEval/Patterns.h"
+
+#include "lib/Dialect/Polynomial/IR/PolynomialOps.h"
+#include "llvm/include/llvm/ADT/TypeSwitch.h"           // from @llvm-project
+#include "llvm/include/llvm/Support/Debug.h"            // from @llvm-project
+#include "mlir/include/mlir/Dialect/Arith/IR/Arith.h"   // from @llvm-project
+#include "mlir/include/mlir/IR/ImplicitLocOpBuilder.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/Passes.h"        // from @llvm-project
+#include "mlir/include/mlir/Transforms/WalkPatternRewriteDriver.h"  // from @llvm-project
+
+#define DEBUG_TYPE "lower-polynomial-eval"
+
+namespace mlir {
+namespace heir {
+
+using polynomial::EvalOp;
+using polynomial::TypedFloatPolynomialAttr;
+
+TypedAttr getScalarOrDenseAttr(Type tensorOrScalarType, APFloat value) {
+  return TypeSwitch<Type, TypedAttr>(tensorOrScalarType)
+      .Case<FloatType>(
+          [&](FloatType type) { return FloatAttr::get(type, value); })
+      .Case<ShapedType>(
+          [&](ShapedType type) { return DenseElementsAttr::get(type, value); })
+      .Default([](Type) { return nullptr; });
+}
+
+LogicalResult LowerViaHorner::matchAndRewrite(EvalOp op,
+                                              PatternRewriter& rewriter) const {
+  Type evaluatedType = op.getValue().getType();
+  ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+  b.setInsertionPoint(op);
+
+  LLVM_DEBUG(llvm::dbgs() << "evaluatedType: " << evaluatedType << "\n");
+
+  auto attr =
+      dyn_cast<polynomial::TypedFloatPolynomialAttr>(op.getPolynomialAttr());
```

Comment:
remind me, do we always expect poly.eval from the poly approx patterns to generated a typed float polynomial attr? (the verifier also allows int, but also "FloatPolynomialAttr") - could you add a quick comment at the return failure() below?

or maybe the intention was that another pattern will later match on other poly types?

Reply:
I think we can generalize later to include those options, but right now the polynomial approximation pass always produces a typed float polynomial attr